server/src/test/java/org/elasticsearch/cluster/metadata/DesiredNodeTests.java

```diff
@@ -225,16 +226,63 @@ public void testDesiredNodeIsCompatible() {
-    }
+    public void testEqualsOrProcessorsCloseTo() {
+        final Settings settings = Settings.builder().put(NODE_NAME_SETTING.getKey(), randomAlphaOfLength(10)).build();
+        final double maxDelta = 1E-3;
+
+        final double processorCount = randomNumberOfProcessors();
+        final boolean isEqualOrCloseTo = randomBoolean();
+        final ByteSizeValue memory = ByteSizeValue.ofGb(randomIntBetween(1, 32));
+        final ByteSizeValue storage = ByteSizeValue.ofGb(randomIntBetween(128, 256));
+
+        final DesiredNode desiredNode1;
+        final DesiredNode desiredNode2;
+        if (randomBoolean()) {
+            desiredNode1 = new DesiredNode(settings, processorCount, memory, storage, Version.CURRENT);
+            desiredNode2 = new DesiredNode(
+                settings,
+                isEqualOrCloseTo ? (float) processorCount : processorCount + maxDelta,
+                memory,
+                storage,
+                Version.CURRENT
+            );
+        } else {
+            final double desiredNodes1Min = processorCount;
+            final Double desiredNodes1Max = randomBoolean() ? processorCount + randomIntBetween(1, 10) : null;
+            final DesiredNode.ProcessorsRange desiredNodes1ProcessorsRange = new DesiredNode.ProcessorsRange(
+                desiredNodes1Min,
+                desiredNodes1Max
+            );
 
-    private float randomFloatBetween(float start, float end) {
-        float result = 0.0f;
-        while (result < start || result > end || Float.isNaN(result)) {
-            result = start + randomFloat() * (end - start);
+            final double modifiedMinProcessors = isEqualOrCloseTo ? (float) desiredNodes1Min : desiredNodes1Min + maxDelta;
+
+            final double desiredNodes2Min;
+            final Double desiredNodes2Max;
+            if (desiredNodes1Max != null && randomBoolean()) {
+                desiredNodes2Min = randomBoolean() ? desiredNodes1Min : modifiedMinProcessors;
+                desiredNodes2Max = isEqualOrCloseTo ? desiredNodes1Max.floatValue() : desiredNodes1Max + maxDelta;
+            } else {
+                desiredNodes2Min = modifiedMinProcessors;
+                desiredNodes2Max = desiredNodes1Max;
+            }
+            final DesiredNode.ProcessorsRange desiredNodes2ProcessorsRange = new DesiredNode.ProcessorsRange(
+                desiredNodes2Min,
+                desiredNodes2Max
+            );
+
+            desiredNode1 = new DesiredNode(settings, desiredNodes1ProcessorsRange, memory, storage, Version.CURRENT);
+            desiredNode2 = new DesiredNode(settings, desiredNodes2ProcessorsRange, memory, storage, Version.CURRENT);
         }
 
-        return result;
+        assertThat(desiredNode1.equalsWithProcessorsCloseTo(desiredNode2), is(isEqualOrCloseTo));
+    }
+
+    private double randomNumberOfProcessors() {
+        return randomDoubleBetween(Math.ulp(0.0), 512.99999999, true);
```

Comment:
Is `Math.ulp(0.0)` not the same as `Double.MIN_VALUE`? I'd prefer to use the latter then.